src/GafferArnold/IECoreArnoldPreview/Renderer.cpp

```diff
@@ -3465,45 +3465,80 @@ class ArnoldRenderer final : public ArnoldRendererBase
 
 		ArnoldRenderer( RenderType renderType, const std::string &fileName, const IECore::MessageHandlerPtr &messageHandler )
 			:	ArnoldRendererBase( nodeDeleter( renderType ) ),
-				m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
+				m_messageHandler( messageHandler ), m_globals( new ArnoldGlobals( renderType, fileName, m_shaderCache.get() ) )
 		{
+			if( m_messageHandler )
+			{
+				AiMsgSetCallback( &ArnoldRenderer::aiMsgCallback );
+			}
 		}
 
 		~ArnoldRenderer() override
 		{
 			pause();
+			if( g_currentMessageHandler == m_messageHandler )
+			{
+				g_currentMessageHandler = nullptr;
+			}
```

Comment:
Shouldn't we be calling `AiMsgResetCallback()` here? Otherwise, I think we'll lose the console output from a handler-less render after perform a  handlered render.

I think the `if( g_currentMessageHandler == m_messageHandler )` check is also probably misleading? We can only have one `ArnoldRenderer` active at any given time, so no other condition is possible. If we could have more than one renderer, the check still wouldn't provide security, because there's nothing to prevent two renderers being given the same handler. Perhaps, if anything, `assert( g_currentMessageHandler == m_messageHandler )` is more appropriate.

Reply:
Refactored in https://github.com/GafferHQ/gaffer/pull/3767/commits/d917f02a78e2c4a5bc1f11efa98d8fed04f9c996